File: sdk/go/keygen.go
package psnx

import (
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"io"
	"math/big"

	"golang.org/x/crypto/hkdf"
	"golang.org/x/crypto/pbkdf2"
)

// KeyPair represents a vault key pair with public key
type KeyPair struct {
	VaultKey    []byte
	PublicKey   *big.Int
	Fingerprint string
	EntropyBits int
}

// KeyGenerator generates cryptographic keys for the vault
type KeyGenerator struct {
	extraEntropy []byte
}

// NewKeyGenerator creates a new key generator
func NewKeyGenerator() *KeyGenerator {
	return &KeyGenerator{}
}

// NewKeyGeneratorWithEntropy creates a key generator with extra entropy
func NewKeyGeneratorWithEntropy(entropy []byte) *KeyGenerator {
	return &KeyGenerator{
		extraEntropy: entropy,
	}
}

// Generate creates a new random vault key
func (kg *KeyGenerator) Generate() ([]byte, error) {
	entropy := make([]byte, 64+len(kg.extraEntropy))
	if _, err := io.ReadFull(rand.Reader, entropy[:64]); err != nil {
		return nil, err
	}
	copy(entropy[64:], kg.extraEntropy)

	// Derive key using HKDF
	hkdfReader := hkdf.New(sha256.New, entropy, []byte("PSNX_KEYGEN_v1"), []byte("vault_master_key"))
	key := make([]byte, KeySize)
	if _, err := io.ReadFull(hkdfReader, key); err != nil {
		return nil, err
	}

	return key, nil
}

// FromPassword derives a vault key from a password
func (kg *KeyGenerator) FromPassword(password string, salt []byte) (vaultKey []byte, usedSalt []byte, err error) {
	if len(password) < 8 {
		return nil, nil, errors.New("password must be at least 8 characters")
	}

	if salt == nil {
		salt = make([]byte, 16)
		if _, err := io.ReadFull(rand.Reader, salt); err != nil {
			return nil, nil, err
		}
	}

	// PBKDF2 with 100,000 iterations
	key := pbkdf2.Key([]byte(password), salt, 100000, KeySize, sha256.New)

	return key, salt, nil
}

// FromEntropy derives a vault key from external entropy
func (kg *KeyGenerator) FromEntropy(entropy []byte, minBits int) ([]byte, error) {
	if len(entropy)*8 < minBits {
		return nil, errors.New("entropy must be at least the minimum bits")
	}

	combined := make([]byte, len(entropy)+32+len(kg.extraEntropy))
	copy(combined, entropy)

	randomPart := make([]byte, 32)
	if _, err := io.ReadFull(rand.Reader, randomPart); err != nil {
		return nil, err
	}
	copy(combined[len(entropy):], randomPart)
	copy(combined[len(entropy)+32:], kg.extraEntropy)

	// Derive key using HKDF
	hkdfReader := hkdf.New(sha256.New, combined, []byte("PSNX_EXTERNAL_ENTROPY"), []byte("vault_key_from_external"))
	key := make([]byte, KeySize)
	if _, err := io.ReadFull(hkdfReader, key); err != nil {
		return nil, err
	}

	return key, nil
}

// GenerateKeyPair generates a complete key pair
func (kg *KeyGenerator) GenerateKeyPair() (*KeyPair, error) {
	vaultKey, err := kg.Generate()
	if err != nil {
		return nil, err
	}

	// Derive ZKP private key
	combined := append([]byte("PSNX_ZKP_PRIVATE_"), vaultKey...)
	hash := sha256.Sum256(combined)

	// Public key calculation (simplified)
	x := new(big.Int).SetBytes(hash[:])
	x.Mod(x, Q)
	if x.Sign() == 0 {
		x.SetInt64(1)
	}

	publicKey := new(big.Int).Exp(G, x, P)

	// Fingerprint
	fpHash := sha256.Sum256(vaultKey)
	fingerprint := bytesToHex(fpHash[:8])

	return &KeyPair{
		VaultKey:    vaultKey,
		PublicKey:   publicKey,
		Fingerprint: fingerprint,
		EntropyBits: 256,
	}, nil
}

// Verify checks if a vault key is valid
func Verify(vaultKey []byte) bool {
	if len(vaultKey) != 32 {
		return false
	}
	// Check it's not all zeros
	for _, b := range vaultKey {
		if b != 0 {
			return true
		}
	}
	return false
}

File: sdk/rust/src/keygen.rs
//! Key generation module

use hkdf::Hkdf;
use num_bigint::BigUint;
use rand::{rngs::OsRng, RngCore};
use sha2::{Digest, Sha256};

use crate::{Error, Result};
use crate::zkp::ZKP_PARAMS;

/// Key pair with vault key and ZKP public key
#[derive(Debug, Clone)]
pub struct KeyPair {
    pub vault_key: [u8; 32],
    pub public_key: BigUint,
    pub fingerprint: String,
    pub entropy_bits: u32,
}

/// Key generator
pub struct KeyGenerator {
    extra_entropy: Vec<u8>,
}

impl KeyGenerator {
    /// Create a new key generator
    pub fn new() -> Self {
        Self {
            extra_entropy: Vec::new(),
        }
    }
    
    /// Create with extra entropy
    pub fn with_entropy(entropy: &[u8]) -> Self {
        Self {
            extra_entropy: entropy.to_vec(),
        }
    }
    
    /// Generate a random vault key
    pub fn generate(&self) -> [u8; 32] {
        let mut entropy = vec![0u8; 64];
        OsRng.fill_bytes(&mut entropy);
        entropy.extend_from_slice(&self.extra_entropy);
        
        let hkdf = Hkdf::<Sha256>::new(Some(b"PSNX_KEYGEN_v1"), &entropy);
        let mut vault_key = [0u8; 32];
        hkdf.expand(b"vault_master_key", &mut vault_key).unwrap();
        
        vault_key
    }
    
    /// Derive key from password using Scrypt-like KDF
    pub fn from_password(password: &str, salt: Option<&[u8]>) -> Result<([u8; 32], Vec<u8>)> {
        if password.len() < 8 {
            return Err(Error::Validation(
                "Password must be at least 8 characters".to_string()
            ));
        }
        
        let salt = salt.map(|s| s.to_vec()).unwrap_or_else(|| {
            let mut s = vec![0u8; 16];
            OsRng.fill_bytes(&mut s);
            s
        });
        
        // PBKDF2-like derivation (simplified)
        let hkdf = Hkdf::<Sha256>::new(Some(&salt), password.as_bytes());
        let mut vault_key = [0u8; 32];
        hkdf.expand(b"vault_key_from_password", &mut vault_key)
            .map_err(|e| Error::Key(e.to_string()))?;
        
        Ok((vault_key, salt))
    }
    
    /// Generate from external entropy
    pub fn from_entropy(&self, entropy: &[u8], min_bits: usize) -> Result<[u8; 32]> {
        if entropy.len() * 8 < min_bits {
            return Err(Error::Validation(format!(
                "Entropy must be at least {} bits",
                min_bits
            )));
        }
        
        let mut combined = entropy.to_vec();
        let mut system_entropy = vec![0u8; 32];
        OsRng.fill_bytes(&mut system_entropy);
        combined.extend_from_slice(&system_entropy);
        combined.extend_from_slice(&self.extra_entropy);
        
        let hkdf = Hkdf::<Sha256>::new(Some(b"PSNX_EXTERNAL_ENTROPY"), &combined);
        let mut vault_key = [0u8; 32];
        hkdf.expand(b"vault_key_from_external", &mut vault_key)
            .map_err(|e| Error::Key(e.to_string()))?;
        
        Ok(vault_key)
    }
    
    /// Generate a complete key pair
    pub fn generate_keypair(&self) -> KeyPair {
        let vault_key = self.generate();
        
        // Derive ZKP private key
        let mut hasher = Sha256::new();
        hasher.update(b"PSNX_ZKP_PRIVATE_");
        hasher.update(&vault_key);
        let hash = hasher.finalize();
        
        let x = BigUint::from_bytes_be(&hash) % &ZKP_PARAMS.q;
        let x = if x == BigUint::ZERO {
            BigUint::from(1u32)
        } else {
            x
        };
        
        // Public key
        let public_key = ZKP_PARAMS.g.modpow(&x, &ZKP_PARAMS.p);
        
        // Fingerprint
        let fingerprint = hex::encode(&Sha256::digest(&vault_key)[..8]);
        
        KeyPair {
            vault_key,
            public_key,
            fingerprint,
            entropy_bits: 256,
        }
    }
    
    /// Verify a key is valid
    pub fn verify(vault_key: &[u8]) -> bool {
        vault_key.len() == 32 && !vault_key.iter().all(|&b| b == 0)
    }
}

impl Default for KeyGenerator {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_generate() {
        let gen = KeyGenerator::new();
        let key = gen.generate();
        assert!(KeyGenerator::verify(&key));
    }
    
    #[test]
    fn test_keypair() {
        let gen = KeyGenerator::new();
        let keypair = gen.generate_keypair();
        assert!(KeyGenerator::verify(&keypair.vault_key));
        assert!(!keypair.fingerprint.is_empty());
    }
}
